src/txreconciliation.cpp

```diff
@@ -0,0 +1,893 @@
+     * Since it may happen after a delay (at least network latency), new transactions may come
+     * during that time. To avoid mixing old and new transactions, those which are subject for
+     * extension of a current reconciliation round are moved to a reconciliation set snapshot
+     * after an initial (non-extended) sketch is sent.
+     * New transactions are kept in the regular reconciliation set.
+     */
+    std::set<uint256> m_local_set_snapshot;
+
+    /**
+     * A reconciliation round may involve an extension, in which case we should remember
+     * a capacity of the sketch sent out initially, so that a sketch extension is of the same size.
+     */
+    uint16_t m_capacity_snapshot{0};
+
+    /**
+     * Reconciliation sketches are computed over short transaction IDs.
+     * This is a cache of these IDs enabling faster lookups of full wtxids,
+     * useful when peer will ask for missing transactions by short IDs
+     * at the end of a reconciliation round.
+     */
+    std::map<uint32_t, uint256> m_local_short_id_mapping;
+
+    /** Keep track of reconciliations with the peer. */
+    ReconciliationInitByUs m_state_init_by_us;
+    ReconciliationInitByThem m_state_init_by_them;
+
+    /**
+     * Reconciliation sketches are computed over short transaction IDs.
+     * Short IDs are salted with a link-specific constant value.
+     */
+    uint32_t ComputeShortID(const uint256 wtxid) const
+    {
+        const uint64_t s = SipHashUint256(m_k0, m_k1, wtxid);
+        const uint32_t short_txid = 1 + (s & 0xFFFFFFFF);
+        return short_txid;
+    }
+
+    /**
+     * Reconciliation involves computing a space-efficient representation of transaction identifiers
+     * (a sketch). A sketch has a capacity meaning it allows reconciling at most a certain number
+     * of elements (see BIP-330).
+     */
+    Minisketch ComputeSketch(uint16_t capacity, bool use_snapshot=false)
+    {
+        Minisketch sketch;
+        std::set<uint256> working_set;
+
+        if (use_snapshot) {
+            working_set = m_local_set_snapshot;
+        } else {
+            working_set = m_local_set;
+            m_capacity_snapshot = capacity;
+        }
+        // Avoid serializing/sending an empty sketch.
+        if (working_set.size() == 0 || capacity == 0) return sketch;
+
+        std::vector<uint32_t> short_ids;
+        for (const auto& wtxid: working_set) {
+            uint32_t short_txid = ComputeShortID(wtxid);
+            short_ids.push_back(short_txid);
+            m_local_short_id_mapping.emplace(short_txid, wtxid);
+        }
+
+        capacity = std::min(capacity, MAX_SKETCH_CAPACITY);
+        sketch = Minisketch(RECON_FIELD_SIZE, 0, capacity);
+        if (sketch) {
+            for (const uint32_t short_id: short_ids) {
+                sketch.Add(short_id);
+            }
+        }
+        return sketch;
+    }
+
+    public:
```

Comment:
Do you need to make those `ReconciliationState` methods public ?

Reply:
Currently, nothing has to be made public, because it's a friend class. So you think everything should be private?

Reply:
Yes I think it should be private by default. In the future, if a use-case arise, we can make them public.